src/gwt/src/org/rstudio/studio/client/rsconnect/ui/RSConnectDeploy.java

```diff
@@ -830,12 +831,41 @@ private void populateDeploymentFiles(final ProgressIndicator indicator)
       {
          if (StringUtil.isNullOrEmpty(source_.getDeployFile()))
          {
-            indicator.onError(
-                  "Only rendered documents can be published to RStudio Connect. " +
-                  "To publish this document, click Knit to render it, then click " +
-                  "the Publish button above the rendered document.");
-            indicator.onCompleted(); 
-            return; 
+            // Want to make sure the dialog (e.g. publish wizard) has gone away before the message
+            // is shown. In some scenarios, varying both by dialog and by platform (desktop vs.
+            // server), the message is showing while the dialog is still visible but not fully
+            // initialized. In another scenario, the dialog does not dismiss when
+            // indicator.onCompleted() is invoked unless we delay the invocation a bit. This is
+            // all rather unsatisfying.
+            Timer onCompletedTimer = new Timer()
+            {
+               @Override
+               public void run()
+               {
+                  indicator.onCompleted();
+         
+                  // Want to make sure the publish wizard has gone away before this message is shown.
+                  // On Desktop (macOS, at least), trying to do this with indicator.onError(msg)
+                  // leaves the publish wizard visible in an incomplete state and it doesn't dismiss
+                  // until the message is dismissed, so defer message display a bit.
+                  Timer showMessageTimer = new Timer()
+                  {
+                     @Override
+                     public void run()
+                     {
+                        RStudioGinjector.INSTANCE.getGlobalDisplay().showMessage(
+                              GlobalDisplay.MSG_INFO,
+                              "Finished Document Not Found",
+                              "To publish finished document to RStudio Connect, you must first render " +
+                                    "it. Dismiss this message, click Knit to render the document, " +
+                                    "then try publishing again.");
+                     }
+                  };
+                  showMessageTimer.schedule(100);
```

Comment:
Would deferring these to the next event loop work? (i.e. `Scheduler.ScheduleDeferred`)

Reply:
I tried that before I switched to a timer for the initial case(i.e. the inner timer) and it did not help.

I didn't try it on the second case (that prompted the second timer). I'll try that and see if I can eliminate one of the timers at least.